src/pyqasm/pulse/openpulse_visitor.py

```diff
@@ -0,0 +1,997 @@
+            return_value, _ = self._visit_function_call(r_value)
+            _id = l_value.identifier  # type: ignore
+            _, _type, _ = self._get_identifier(statement, _id)
+            if _type:
+                self._qasm_visitor.visit_statement(statement)
+            if _type is not None:
+                if (f_name in ["get_phase"] and not isinstance(_type, pulse_ast.AngleType)) or (
+                    f_name in ["get_frequency"] and not isinstance(_type, pulse_ast.FloatType)
+                ):
+                    raise_qasm3_error(
+                        f"Invalid type '{_type}' for identifier '{_id.name}'",
+                        error_node=r_value,
+                        span=r_value.span,
+                    )
+            statement.rvalue = return_value if return_value is not None else statement.rvalue
+        else:
+            self._qasm_visitor.visit_statement(statement)
+
+        return [statement]
+
+    def _visit_classical_declaration(  # pylint: disable=too-many-branches, too-many-statements, too-many-locals
+        self, statement: pulse_ast.ClassicalDeclaration
+    ) -> list[pulse_ast.Statement]:
+        """Visit a classical operation element.
+
+        Args:
+            statement (ClassicalType): The classical operation to visit.
+
+        Returns:
+            None
+        """
+
+        if isinstance(statement.type, pulse_ast.PortType):
+            if statement.identifier:
+                _port_name = statement.identifier.name
+                if _port_name in self._port_to_qubit_map:
+                    raise_qasm3_error(
+                        f"Port '{_port_name}' already declared",
+                        error_node=statement,
+                        span=statement.span,
+                    )
+                self._port_to_qubit_map[_port_name] = []
+
+        elif isinstance(statement.type, pulse_ast.FrameType):
+            _frame_name = ""
+            freq_arg_value = 0.0
+            freq_arg_type = qasm3_ast.FloatType(qasm3_ast.IntegerLiteral(32))
+            phase_arg_value = 0.0
+            phase_arg_type = qasm3_ast.AngleType(qasm3_ast.IntegerLiteral(32))
+            time_arg_value = qasm3_ast.DurationLiteral(
+                0,
+                unit=(
+                    qasm3_ast.TimeUnit.dt
+                    if self._module._device_cycle_time
+                    else qasm3_ast.TimeUnit.ns
+                ),
+            )
+            if statement.identifier:
+                _frame_name = statement.identifier.name
+                if _frame_name in self._frames:
+                    raise_qasm3_error(
+                        f"Frame '{_frame_name}' already declared",
+                        error_node=statement,
+                        span=statement.span,
+                    )
+
+            if statement.init_expression:
+                _frame_args = statement.init_expression.arguments  # type: ignore
+                if not 3 <= len(_frame_args) <= 4:
```

Comment:
Again, can we refactor validations to the `pulse/validator`? Should decrease the LOC here inside the visitor